src/test/java/io/cryostat/core/util/RuleFilterParserTest.java

```diff
@@ -0,0 +1,170 @@
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.concurrent.RunnableFuture;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.openjdk.jmc.common.item.IItemCollection;
+import org.openjdk.jmc.common.util.IPreferenceValueProvider;
+import org.openjdk.jmc.common.util.TypedPreference;
+import org.openjdk.jmc.flightrecorder.rules.IRule;
+import org.openjdk.jmc.flightrecorder.rules.Result;
+
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public class RuleFilterParserTest {
+
+    RuleFilterParser parser;
+    @Mock Collection<IRule> iRules;
+    Set<String> ruleSet;
+    Set<String> topicSet;
+    TestRule rule0, rule1, rule2, rule3, rule4;
+
+    @BeforeEach
+    void setup() {
+        rule0 = new TestRule("Rule0", "Topic0");
+        rule1 = new TestRule("Rule1", "Topic1");
+        rule2 = new TestRule("Rule2", "Topic2");
+        rule3 = new TestRule("Rule3", "Topic0");
+        rule4 = new TestRule("Rule4", "Topic0");
+
+        ruleSet = Set.of("Rule0", "Rule1", "Rule2", "Rule3", "Rule4");
+        topicSet = Set.of("Topic0", "Topic1", "Topic2");
+
+        Mockito.when(iRules.stream()).thenReturn(Stream.of(rule0, rule1, rule2, rule3, rule4));
+
+        parser = new RuleFilterParser(ruleSet, topicSet);
+    }
+
+    @Test
+    void shouldAcceptAllRules() {
+        String rawFilter = "Rule0,Rule1, rule4,     rule2, rule3   ";
+
+        Predicate<IRule> result = parser.parse(rawFilter);
+        Collection<IRule> rules = iRules.stream().filter(result).collect(Collectors.toList());
+
+        MatcherAssert.assertThat(result, Matchers.notNullValue());
+        MatcherAssert.assertThat(rules, Matchers.hasSize(5));
+    }
+
+    @Test
+    void shouldAcceptSomeRulesofTopic() {
+        String rawFilter = " topic0 ";
+
+        Predicate<IRule> result = parser.parse(rawFilter);
+        Collection<IRule> rules = iRules.stream().filter(result).collect(Collectors.toList());
+
+        for (var e : rules) {
+            System.out.println(e.getId());
+        }
+
+        MatcherAssert.assertThat(result, Matchers.notNullValue());
+        MatcherAssert.assertThat(rules, Matchers.hasSize(3));
+    }
+
+    @Test
+    void shouldAcceptAllWhenFilterNull() {
+        String rawFilter = null;
+
+        Predicate<IRule> result = parser.parse(rawFilter);
+        Collection<IRule> rules = iRules.stream().filter(result).collect(Collectors.toList());
+
+        MatcherAssert.assertThat(result, Matchers.notNullValue());
+        MatcherAssert.assertThat(rules, Matchers.hasSize(5));
+    }
+
+    @Test
+    void shouldAcceptAllWhenFilterBlank() {
```

Comment:
You could combine this test with the last one using `@ParameterizedTest @NullAndEmptySource`. It's not much of a savings, and the test intent is already crystal clear the way you've written it, so this is not really a suggestion to change. Just a neat trick to take a look at. This one taken from Cryostat's `RuleTest`.

```java

    @ParameterizedTest
    @NullAndEmptySource
    void shouldThrowOnBlankName(String s) {
        IllegalArgumentException ex =
                Assertions.assertThrows(
                        IllegalArgumentException.class,
                        () -> {
                            builder.name(s)
                                    .matchExpression(MATCH_EXPRESSION)
                                    .eventSpecifier(EVENT_SPECIFIER)
                                    .build();
                        });
        MatcherAssert.assertThat(
                ex.getMessage(),
                Matchers.containsString("\"name\" cannot be blank, was \"" + s + "\""));
    }
```